x/mock/simulation/random_simulate_blocks.go

```diff
@@ -39,43 +43,162 @@ func SimulateFromSeed(
 		events[what]++
 	}
 
-	app.InitChain(abci.RequestInitChain{AppStateBytes: appStateFn(r, addrs)})
+	time := int64(0)
+	timeDiff := maxTimePerBlock - minTimePerBlock
+
+	res := app.InitChain(abci.RequestInitChain{AppStateBytes: appStateFn(r, keys, accs)})
+	validators := make(map[string]mockValidator)
+	for _, validator := range res.Validators {
+		validators[string(validator.Address)] = mockValidator{validator, GetMemberOfInitialState(r, initialLivenessWeightings)}
+	}
+
 	for i := 0; i < len(setups); i++ {
 		setups[i](r, keys)
 	}
-	app.Commit()
 
-	header := abci.Header{Height: 0}
+	header := abci.Header{Height: 0, Time: time}
+	opCount := 0
+
+	request := abci.RequestBeginBlock{Header: header}
+
+	var pastTimes []int64
 
 	for i := 0; i < numBlocks; i++ {
-		app.BeginBlock(abci.RequestBeginBlock{})
 
-		// Make sure invariants hold at beginning of block and when nothing was
-		// done.
+		// Log the header time for future lookup
+		pastTimes = append(pastTimes, header.Time)
+
+		// Run the BeginBlock handler
+		app.BeginBlock(request)
+
+		// Make sure invariants hold at beginning of block
 		AssertAllInvariants(t, app, invariants, log)
 
 		ctx := app.NewContext(false, header)
 
-		// TODO: Add modes to simulate "no load", "medium load", and
-		// "high load" blocks.
-		for j := 0; j < blockSize; j++ {
+		var thisBlockSize int
+		load := r.Float64()
+		switch {
+		case load < 0.33:
+			thisBlockSize = 0
+		case load < 0.66:
+			thisBlockSize = r.Intn(blockSize * 2)
+		default:
+			thisBlockSize = r.Intn(blockSize * 4)
+		}
+		for j := 0; j < thisBlockSize; j++ {
 			logUpdate, err := ops[r.Intn(len(ops))](t, r, app, ctx, keys, log, event)
 			log += "\n" + logUpdate
 
 			require.Nil(t, err, log)
-			AssertAllInvariants(t, app, invariants, log)
+			if onOperation {
+				AssertAllInvariants(t, app, invariants, log)
+			}
+			fmt.Printf("\rSimulating... block %d/%d, operation %d.", header.Height, numBlocks, opCount)
```

Comment:
I imagine the intended purpose of this is to gauge how long its taking. If so, perhaps add a time field here?